lib/handlers/authenticate-handler.js

```diff
@@ -0,0 +1,255 @@
+var Promise = require('bluebird');
+var Request = require('../request');
+var Response = require('../response');
+var ServerError = require('../errors/server-error');
+var UnauthorizedRequestError = require('../errors/unauthorized-request-error');
+
+/**
+ * Constructor.
+ */
+
+function AuthenticateHandler(options) {
+  options = options || {};
+
+  if (!options.model) {
+    throw new InvalidArgumentError('Missing parameter: `model`');
+  }
+
+  if (!options.model.getAccessToken) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `getAccessToken()`');
+  }
+
+  if (options.scope && undefined === options.addAcceptedScopesHeader) {
+    throw new InvalidArgumentError('Missing parameter: `addAcceptedScopesHeader`');
+  }
+
+  if (options.scope && undefined === options.addAuthorizedScopesHeader) {
+    throw new InvalidArgumentError('Missing parameter: `addAuthorizedScopesHeader`');
+  }
+
+  if (options.scope && !options.model.validateScope) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `validateScope()`');
+  }
+
+  this.addAcceptedScopesHeader = options.addAcceptedScopesHeader;
+  this.addAuthorizedScopesHeader = options.addAuthorizedScopesHeader;
+  this.model = options.model;
+  this.scope = options.scope;
+}
+
+/**
+ * Authenticate Handler.
+ */
+
+AuthenticateHandler.prototype.handle = function(request, response) {
+  if (!(request instanceof Request)) {
+    throw new InvalidArgumentError('Invalid argument: `request` must be an instance of Request');
+  }
+
+  if (!(response instanceof Response)) {
+    throw new InvalidArgumentError('Invalid argument: `response` must be an instance of Request');
+  }
+
+  return Promise.bind(this)
+    .then(function() {
+      return this.getToken(request);
+    })
+    .then(function(token) {
+      return this.getAccessToken(token);
+    })
+    .tap(function(token) {
+      return this.validateAccessToken(token);
+    })
+    .tap(function(token) {
+      if (!this.scope) {
+        return;
+      }
+
+      return this.validateScope(token);
+    })
+    .tap(function(token) {
+      return this.updateResponse(response, token);
+    })
+    .catch(function(e) {
+      // Include the "WWW-Authenticate" response header field if the client
+      // lacks any authentication information.
+      //
+      // @see https://tools.ietf.org/html/rfc6750#section-3.1
+      if (e instanceof UnauthorizedRequestError) {
+        response.set('WWW-Authenticate', 'Bearer realm="Service"');
+      }
+
+      if (!(e instanceof OAuthError)) {
+        throw new ServerError(e);
+      }
+
+      throw e;
+    });
+};
+
+/**
+ * Get the token from the header or body, depending on the request.
+ *
+ * "Clients MUST NOT use more than one method to transmit the token in each request."
+ *
+ * @see https://tools.ietf.org/html/rfc6750#section-2
+ */
+
+AuthenticateHandler.prototype.getToken = function(request) {
```

Comment:
`getToken` as a name makes it seem like it's retrieving from the model.  Something like `getTokenFromRequest` would be better.